cypress/events/index.js

```diff
@@ -186,41 +188,82 @@ module.exports = function setupNodeEvents (on, config) {
     }
 
     return new Promise((resolve) => {
       const pkgFilePath = pathToPackageFile(plugin)
-      console.log(`Waiting for ${pkgFilePath} to exist`)
+      log(`Waiting for ${pkgFilePath} to exist`)
       return existsFile(pkgFilePath, timeout).then(() => {
         if (version) {
           const packageContent = fs.readFileSync(pkgFilePath, 'utf8')
           const { version: currentVersion } = JSON.parse(packageContent)
-          console.log(`Current version of ${plugin} is @${currentVersion}`)
+          log(`Current version of ${plugin} is @${currentVersion}`)
           if (version === '@' + currentVersion) {
             resolve(makeSureCypressCanInterpretTheResult)
           } else {
             retry().then(() => resolve(makeSureCypressCanInterpretTheResult))
           }
         } else {
-          console.log('Skip version test')
+          log('Skip version test')
           resolve(makeSureCypressCanInterpretTheResult)
         }
       })
     })
   }
 
-  const restoreStarterFiles = () => {
-    const { projectFolder } = config.env
-    const appDir = path.join(projectFolder, 'app')
-    return fse.emptyDir(appDir)
-      .then(() => fse.copy(starterDir, projectFolder))
-      .then(() => waitUntilAppRestarts())
+  const backupStarterFiles = () => {
+    const projectDir = path.join(config.env.projectFolder)
+    const backupDir = path.join(config.env.tempFolder, 'backupStarterFiles')
+
+    // Define the filter function
+    const filter = (dir) => !dir.includes('node_modules') && !dir.includes('package-lock.json')
+
+    return fse.emptyDir(backupDir)
+      // Copy the files using the filter
+      .then(() => fse.copy(projectDir, backupDir, { filter }))
       .then(makeSureCypressCanInterpretTheResult)
   }
 
+  const restoreStarterFiles = async (remainingRetries = 4) => {
+    try {
+      const tmpDir = path.join(config.env.projectFolder, '.tmp')
+      const appDir = path.join(config.env.projectFolder, 'app')
+      const backupDir = path.join(config.env.tempFolder, 'backupStarterFiles')
+      const projectDir = path.join(config.env.projectFolder)
+
+      const originalPackageJsonHash = await getFileHash(path.join(backupDir, 'package.json'))
+      const currentPackageJsonHash = await getFileHash(path.join(projectDir, 'package.json'))
+
+      // Copy the files
+      await fse.emptyDir(tmpDir)
+      await fse.emptyDir(appDir)
+      await fse.copy(backupDir, projectDir)
+      if (originalPackageJsonHash !== currentPackageJsonHash) {
+        log('Restoring to starter plugins')
+        const command = 'npm prune && npm install'
+        await exec(command, { cwd: config.env.projectFolder })
+        await sleep(1000)
+        log(`Completed ${command}`)
+      }
+      await waitUntilAppRestarts()
+      return makeSureCypressCanInterpretTheResult()
+    } catch (error) {
+      if (remainingRetries > 0) {
+        remainingRetries = typeof remainingRetries === 'number' ? remainingRetries - 1 : 0
```

Comment:
Optional change:  It's cleaner not to overwite the parameter, you could move the line 258 e.g.

```
return restoreStarterFiles(typeof remainingRetries === 'number' ? remainingRetries - 1 : 0)
```

Reply:
I'm happy with what I've got so I'll merge